### ISSUE TITLE
Possible gradient explosion when running with default configuration

### ISSUE THREAD

**Issue by sissykosm** [pedrolarben/ElectricDemandForecasting-DL#1]
Greetings @pedrolarben, 

I tried to replicate your results with the same exact datasets and when I ran the provided code, both from `sample_demo.py` and from `main.py`, and after some epochs, the loss outputs erroneous values as displayed on this screenshot.

<img width="960" alt="Screenshot 2021-01-30 at 21 54 40" src="https://user-images.githubusercontent.com/44316099/106367830-c392c080-6345-11eb-9253-3ea303b2647f.png">

From my experience, this can indicate a possible gradient explosion during training. 
Have you got any insight on why is this happening? Have you encountered the same issue when you were performing the experiments? 

I'm using Tensorflow 2.4 and I'm training on an NVIDIA RTX 2080 GPU. 

Thanks a lot in advance :)

**Comment by pedrolarben** [comment]
Hi @sissykosm, thanks for pointing out this possible bug.
I'll check asap what may be causing this behaviour.
Could you give me more details about your execution parameters in order to reproduce the error? 
- Model architecture, model optimizer and loss function:
- Forecast horizon and past history:
- Batch size:
- Normalization method:

**Comment by pedrolarben** [comment]
@sissykosm I have trained a CNN and an LSTM network for more than 50 epochs and there is no sign of gradient explosion.
 
I saw you are using a transformer with a fixed learning rate, right?
I'm not sure what might be causing the problem, but I would give a try to change the optimiser.

In the "Attention Is All You Need" paper [[1]](https://arxiv.org/abs/1706.03762), the authors mention that they vary the learning rate over the course of training.
You can find the authors explanations in this issue https://github.com/tensorflow/tensor2tensor/issues/280. 
Also, it has been deeply studied the importance of the warm-up stage in training transformers, have a look at [[2]](https://openreview.net/pdf?id=B1x8anVFPr).
> Different from the optimization of many other architectures in which the learning rate starts from a relatively large value and then decays, a learning rate warm-up stage for the Post-LN Transformer is critical. [...] After this warm-up stage, the learning rate will be set by classical learning rate schedulers, such as the linear decay, the inverse square-root decay, or forced decay at particular iterations.

**Comment by sissykosm** [comment]
> Hi @sissykosm, thanks for pointing out this possible bug.
> I'll check asap what may be causing this behaviour.
> Could you give me more details about your execution parameters in order to reproduce the error?
> 
> * Model architecture, model optimizer and loss function:
> * Forecast horizon and past history:
> * Batch size:
> * Normalization method:

Hello Pedro, thanks for the reply and sorry for the delay.
I was running your TCN, from experiments.py with:
```
TCN_PARAMS = {'nb_filters': [128],'kernel_size': [6],'nb_stacks': [2],'dilations': [[1, 3, 6, 12, 24]],'dropout_rate': [0],
BATCH_SIZE = [128], epochs = 50
history = 288
```
I noticed that by changing the activation function in the TCN layers from "linear" to "relu" fixed the issue with the exploding gradients.

The tests I did with the Transformer are quite irrelevant to the issue I mentioned (just liked your preprocessing functions/experimental set-up and wanted to test a new model), but the detail you point out on the optimization process with the changing lr is indeed a good addition.

Best,

Sissy
